docs/src/tutorials/nonlinear/user_defined_hessians.jl

```diff
@@ -0,0 +1,210 @@
+# ``V``! However, because ``V`` solves an optimization problem internally, we
+# can't use automatic differentiation to compute the first and second
+# derivatives. Instead, we can use JuMP's ability to pass callback functions
+# for the gradient and Hessian instead.
+
+# First up, we need to define the gradient of ``V`` with respect to ``x``. In
+# general, this may be difficult to compute, but because ``x`` appears only in
+# the objective, we can just differentiate the objective function with respect
+# to ``x``, giving:
+
+function ∇V(g::AbstractVector, x...)
+    _, y = solve_lower_level(x...)
+    g[1] = 2 * x[1] * y[1] - y[1]^4
+    g[2] = 2 * x[2] * y[2] - 2 * y[2]^4
+    return
+end
+
+# Second, we need to define the Hessian of ``V`` with respect to ``x``. This is
+# a symmetric matrix, but in our example only the diagonal elements are
+# non-zero:
+
+function ∇²V(H::AbstractMatrix, x...)
+    _, y = solve_lower_level(x...)
+    H[1, 1] = 2 * y[1]
+    H[2, 2] = 2 * y[2]
+    return
+end
+
+# We now have enough to define our bilevel optimization problem:
+
+model = Model(Ipopt.Optimizer)
+@variable(model, x[1:2] >= 0)
+register(model, :V, 2, V, ∇V, ∇²V)
+@NLobjective(model, Min, x[1]^2 + x[2]^2 + V(x[1], x[2]))
+optimize!(model)
+solution_summary(model)
+
+# The optimal objective value is:
+
+objective_value(model)
+
+# and the optimal upper-level decision variables ``x`` are:
+
+value.(x)
+
+# To compute the optimal lower-level decision variables, we need to call
+# `solve_lower_level` with the optimal upper-level decision variables:
+
+_, y = solve_lower_level(value.(x)...)
+y
+
+# ## Memoization
+
+# Our solution approach works, but it has a performance problem: every time
+# we need to compute the value, gradient, or Hessian of ``V``, we have to
+# re-solve the lower-level optimization problem! This is wasteful, because we
+# will often call the gradient and Hessian at the same point, and so solving the
+# problem twice with the same input repeats work unnecessarily.
+
+# We can work around this by using memoization:
+
+function memoized_solve_lower_level()
+    last_x, f, y = nothing, NaN, [NaN, NaN]
+    function _update_if_needed(x...)
+        if last_x !== x
+            f, y = solve_lower_level(x...)
+            last_x = x
+        end
+        return
+    end
+    function memoized_f(x...)
+        _update_if_needed(x...)
+        return f
+    end
+    function memoized_∇f(g::AbstractVector, x...)
+        _update_if_needed(x...)
+        g[1] = 2 * x[1] * y[1] - y[1]^4
+        g[2] = 2 * x[2] * y[2] - 2 * y[2]^4
+        return
+    end
+    function memoized_∇²f(H::AbstractMatrix, x...)
+        _update_if_needed(x...)
+        H[1, 1] = 2 * y[1]
+        H[2, 2] = 2 * y[2]
+        return
+    end
+    return memoized_f, memoized_∇f, memoized_∇²f
+end
+
+f, ∇f, ∇²f = memoized_solve_lower_level()
+
+# The function above is a little confusing, but it returns three new functions
```

Comment:
To make the function less confusing, what about storing the state in a struct instead of a closure?